### PR TITLE
Add v2 unit test for loss functions and update loss/metrics

### PR DESCRIPTION
**Pull request by am2145**
## Description
This PR adds a unit test for the loss functions defined in `loss.py`. The tests are not exhaustive, but include a test for each loss function available and check for basic functionality such as expected dimensions, reduction, and simple parameters. Some loss function definitions in `loss.py` are modified to address some issues. These are: 
1) BoundedMSELoss is modified to have the lt and gt masks applied to the targets in addition to preds. 
2) MulticlassMCCLoss is modified to have summations such that the end vector is a scalar. This should match the implementation in v1. 
3) BinaryDirichletLoss and MulticlassDirichletLoss have had their forward call modified to pass to DirichletMixin as the loss was previously returning a NoneType.  

`metrics.py` has also been updated to have metrics inherit from loss functions where possible. Specifically SID and Wasserstein metrics now inherit from the corresponding loss functions. The MCCmetric has been replaced with BinaryMCC and MulticlassMCC metrics to reflect the updated loss functions in loss.py as well. BoundedMixin has been updated to apply the lt and gt masks applied to targets as well, matching the change for BoundedMSELoss in `loss.py`

## Questions
1) The changes made to MulticlassMCCLoss and the DirichletLoss functions work, but are they in line with what was intended for the v2 framework?

### CODE REVIEW

**Review by davidegraff — COMMENTED** [review]
beyond the open questions about calls to `super()`, my only comment is to using `torch.testing` in test files

File: chemprop/v2/models/loss.py [hunk 1]
```diff
@@ -257,16 +257,15 @@ def forward(self, preds: Tensor, targets: Tensor, *args) -> Tensor:
...
+        return super(LossFunction, self).forward(preds, y_one_hot, *args)
 
 
 @LossFunctionRegistry.register("multiclass-dirichlet")
 class MulticlassDirichletLoss(LossFunction, DirichletMixin):
     def forward(self, preds: Tensor, targets: Tensor, mask: Tensor, *args) -> Tensor:
         y_one_hot = torch.eye(preds.shape[2], device=preds.device)[targets.long()]
 
-        return super().forward(preds, y_one_hot, mask)
+        return super(LossFunction, self).forward(preds, y_one_hot, mask)
```
[the remark this answers is not in the document]
Comment (am2145) [reply chemprop/v2/models/loss.py]: No reason as it just happened to be what worked first. I have changed this now to use `DirichletMixin.forward` and the result is the same, so we should be good to go here.

**Comment by am2145** [comment]
> beyond the open questions about calls to `super()`, my only comment is to using `torch.testing` in test files

Thank you for reviewing the PR and the feedback. I have refactored the tests to use `torch.testing` as requested and also updated the DirichletLoss functions. I have also rebased on Hao-Wei and Xiaorui's previous PRs that have been merged in v2/dev already. Please let me know if there are any further changes that are required.

File: chemprop/v2/models/loss.py [hunk 2]
```diff
@@ -257,16 +257,16 @@ def forward(self, preds: Tensor, targets: Tensor, *args) -> Tensor:
         n_tasks = targets.shape[1]
         preds = preds.reshape(len(preds), n_tasks, N_CLASSES)
         y_one_hot = torch.eye(N_CLASSES, device=preds.device)[targets.long()]
-
-        return super().forward(preds, y_one_hot)
+        
+        return DirichletMixin.forward(self, preds, y_one_hot, *args)
```
Comment (davidegraff) [inline chemprop/v2/models/loss.py:261]:
```suggestion
        return super().forward(preds, y_one_hot, *args)
```
If you're using the `super`'s method, then it's standard to call it using the `super()` object

File: chemprop/v2/models/loss.py [hunk 3]
```diff
@@ -257,16 +257,16 @@ def forward(self, preds: Tensor, targets: Tensor, *args) -> Tensor:
...
 
 
 @LossFunctionRegistry.register("multiclass-dirichlet")
 class MulticlassDirichletLoss(LossFunction, DirichletMixin):
     def forward(self, preds: Tensor, targets: Tensor, mask: Tensor, *args) -> Tensor:
         y_one_hot = torch.eye(preds.shape[2], device=preds.device)[targets.long()]
 
-        return super().forward(preds, y_one_hot, mask)
+        return DirichletMixin.forward(self, preds, y_one_hot, mask)
```
Comment (davidegraff) [inline chemprop/v2/models/loss.py:269]: see above

File: chemprop/v2/tests/unit/test_loss_functions.py [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+)
+def test_BinaryDirichlet(preds, targets, mask, w_s, w_t, lt_mask, gt_mask, v_kl, expected_loss):
+    """
+    Test on the dirichlet loss function for classification.
+    Note these values were not hand derived, just testing for
+    dimensional consistency.
+    """
+    binary_dirichlet_loss = BinaryDirichletLoss(v_kl = v_kl)
+    loss = binary_dirichlet_loss(preds, targets, mask, w_s, w_t, lt_mask, gt_mask)
+    print(loss)
```
Comment (davidegraff) [inline chemprop/v2/tests/unit/test_loss_functions.py:125]:
```suggestion

```

File: chemprop/v2/tests/unit/test_loss_functions.py [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+    ]
+)
+def test_BinaryDirichlet_wrong_dimensions(preds, targets, mask, w_s, w_t, lt_mask, gt_mask):
+    """
+    Test on the dirichlet loss function for classification
+    for dimension errors.
+    """
+    with pytest.raises(RuntimeError):
+        binary_dirichlet_loss = BinaryDirichletLoss()
+        binary_dirichlet_loss(preds, targets, mask, w_s, w_t, lt_mask, gt_mask)
```
Comment (davidegraff) [inline chemprop/v2/tests/unit/test_loss_functions.py:150]: good test- should we think about catching this and throwing a more descriptive error?

File: chemprop/v2/models/loss.py (hunk 2 above)
Comment (am2145) [reply chemprop/v2/models/loss.py:261]: When I try this, the forward method from `LossFunction` gets called first. This means that line 52 returns an error as the Loss is `NoneType` because it has not gone through DirichletMixin yet. What's the correct way to use `super()` here?

File: chemprop/v2/tests/unit/test_loss_functions.py (hunk 5 above)
Comment (am2145) [reply chemprop/v2/tests/unit/test_loss_functions.py:150]: I adapted this test from v1. Do we want to print a wrong input dimension error?

Comment (davidegraff) [reply chemprop/v2/tests/unit/test_loss_functions.py:150]: we decided to ignore my comment

**Comment by am2145** [comment]
I have swapped the inheritance of the DirichletLoss functions to have `DirichletMixin` first and then have their forward method use `super()` as we discussed in the meeting. I also deleted the print statement and rebased off of v2/dev so everything should now be up to date. Thanks again for all the help.

**Review by davidegraff — APPROVED** [review]
thanks for the great work @am2145 !
